### ISSUE TITLE
Dropdowns should close on Esc press

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#11568]
Scenario:

1.  Open a drop-down (e.g. headings)
2.  Press Esc

Expected: the dropdown is closed. The focus moved back to the editing.

Actual: nothing happened.

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

**Comment by Reinmar** [comment]
Media embed's panel closes. All other stay open.

**Comment by oleq** [comment]
It never worked for headings if opened using mouse, I checked builds as old as v12.0.0. 

It only works when you're navigating using the keyboard because then the dropdown element (panel, items, whatever) is focused, and  the heading's keystroke handler listens to events from the dropdown element.

> Media embed's panel closes. All other stay open.

Because there's an input in the panel that focuses as soon as you open the dropdown. Focused input == keystroke handler embedded in the media feature is able to listen to events from the media dropdown.

To address this issue, we'd need to:

*   either use editor's global keystroke handler, 
    *   which would break encapsulation of UI views: heading dropdown view has no access to the global keystroke handler of the editor instance
*   or figure out another global system for keystrokes. It would need to aggregate different features, though: 
    *   to avoid 50 listeners to the same DOM event,
    *   to avoid collisions, 
        *   what if nested components should respond to Esc? For instance, a dropdown in a balloon: dropdown should close on first Esc, the balloon on the next. 
            *   The current implementation handles this pretty well because each component has its own keystroke handler which works according to the current focus (no collision).
*   (probably the best idea) or we could start focusing dropdowns (also, buttons, etc.) when clicked. Embedded keystroke handlers will begin to work and do their job. a11y will also improve. Example: [bootstrap dropdowns](https://getbootstrap.com/docs/4.0/components/dropdowns/).

**Comment by Reinmar** [comment]
> *   (probably the best idea) or we could start focusing dropdowns (also, buttons, etc.) when clicked. Embedded keystroke handlers will begin to work and do their job. a11y will also improve. Example: [bootstrap dropdowns](https://getbootstrap.com/docs/4.0/components/dropdowns/).

My thoughts too :+1:

**Comment by mlewand** [comment]
So this issue should be fixed by #11838 let's keep it open to make sure that indeed it works for all dropdowns as expected.

**Comment by Comandeer** [comment]
> Expected: the dropdown is closed. The focus moved back to the editing.

Actually, the focus should probably move back to the element that triggered the dropdown ([ARIA Practices](https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-select-only.html) recommends that). Moving back to editing would be really disturbing to the keyboard-only user, e.g. they opened some dropdown by accident and they're forced to refocus the toolbar and move through all buttons once more just to find the right dropdown.
